File: src/freundtech/bukkit/oneslotserver/TickHandler.java
package freundtech.bukkit.oneslotserver;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

import org.bukkit.GameMode;

public class TickHandler implements Runnable {

	private OneSlotServer plugin;
	
	private SimpleDateFormat timeLeft;
	private SimpleDateFormat hoursWait;

	public TickHandler(OneSlotServer plugin) {
		this.plugin = plugin;
		timeLeft = new SimpleDateFormat("mm:ss");
		timeLeft.setTimeZone(TimeZone.getTimeZone("GMT"));
		hoursWait = new SimpleDateFormat("kk:mm");
		hoursWait.setTimeZone(TimeZone.getTimeZone("GMT"));

	}

	@Override
	public void run() {
		if (plugin.activePlayer != null) {
			Date now = new Date();
			Playerinfo player = plugin.activePlayer;

			if (player.timeleft - (now.getTime() / 1000 - player.joinedAt) <= 0) {
				long waitleft = ((24 * 60 * 60) - (now.getTime()/1000 - player.firstJoin)) * 1000;
				Date date = new Date(waitleft);
				String timestring = hoursWait.format(date);
				
				player.player.kickPlayer("Time is up. You can play again in " + timestring + " hours.");
			}
			
			if (player.player.getGameMode() != GameMode.SPECTATOR) {
				long timeleft = (player.timeleft - (now.getTime()/1000 - player.joinedAt)) * 1000;
				Date date = new Date(timeleft);
				
				String timestring = timeLeft.format(date);
				
				player.player.setPlayerListName(timestring + " left");
			}
		}
	}
}

File: src/main/kotlin/com/freundtech/minecraft/oneslotserver/handler/TickHandler.kt
package com.freundtech.minecraft.oneslotserver.handler

import com.freundtech.minecraft.oneslotserver.OneSlotServer
import com.freundtech.minecraft.oneslotserver.extension.oneSlotServer
import com.freundtech.minecraft.oneslotserver.extension.saveToSharedData
import com.freundtech.minecraft.oneslotserver.extension.setSpectator
import com.freundtech.minecraft.oneslotserver.util.*
import org.bukkit.GameMode
import java.util.*

class TickHandler : Runnable {
    private val plugin = OneSlotServer.instance

    override fun run() {
        val now = currentTime()

        var headerMessage = "No player playing"

        plugin.activePlayer?.let {
            if (it.oneSlotServer.timeLeft - (now - it.oneSlotServer.joinedAt) <= 0) {
                val waitLeft = (plugin.pauseTime - (now - it.oneSlotServer.firstJoin))
                val kickMessage = "Time is up. You can play again in ${waitLeft.format(hoursFormat)} hours."
                if (!it.hasPermission(PERMISSION_SPECTATE)) {

                    it.kickPlayer(kickMessage)
                }
                else {
                    it.sendMessage(kickMessage)
                    it.saveToSharedData()
                    it.setSpectator()
                    plugin.activePlayer = null
                }
            }
            else {
                val timeLeft = (it.oneSlotServer.timeLeft - (now - it.oneSlotServer.joinedAt))
                headerMessage = "${timeLeft.format(minutesFormat)} minutes left"
            }
        }

        plugin.server.onlinePlayers.forEach {
            it.playerListHeader = headerMessage
        }
    }
}
